File: src/main/java/com/hiberus/anaya/redmineapi/TimeEntry.java
package com.hiberus.anaya.redmineapi;

import org.json.JSONObject;

import java.io.IOException;
import java.time.LocalDate;
import java.time.YearMonth;
import java.util.List;
import java.util.Objects;

/**
 * 'You spent X hours into an issue with a message' object
 */
public class TimeEntry {

    /* ------------------------- manager ------------------------- */

    private final RedmineManager manager; // an entry is associated to a manager

    static int getIssueId(JSONObject rawEntry) {
        // returns the id from a rawEntry
        return rawEntry.getJSONObject("issue").getInt("id");
    }

    /* ------------------------- data ------------------------- */

    final int id; // the entry id in the database
    /**
     * The issue id
     */
    public final Issue issue;
    private final LocalDate spent_on; // date it was spent
    private double spent = 0; // hours spent
    private String comment = ""; // comment

    private JSONObject original; // the original raw object, for diff purposes

    /* ------------------------- constructors ------------------------- */

    TimeEntry(JSONObject rawEntry, List<Issue> issues, RedmineManager manager) {
        // creates a new entry from a json raw data
        this.manager = manager;
        original = rawEntry;

        id = rawEntry.getInt("id");
        issue = issues.stream().filter(issue -> issue.id == getIssueId(rawEntry)).findFirst().orElseThrow();
        spent_on = LocalDate.parse(rawEntry.getString("spent_on"));
        spent = rawEntry.getDouble("hours");
        comment = rawEntry.optString("comments");
    }

    TimeEntry(Issue issue, LocalDate spent_on, RedmineManager manager) {
        // Creates a new time entry for an existing issue and date
        this.manager = manager;
        this.id = RedmineManager.NONE;
        this.issue = issue;
        this.spent_on = spent_on;
    }

    /* ------------------------- properties getters ------------------------- */

    /**
     * Checks if this entry was spent on a specific date
     *
     * @param date check spent with this date
     * @return true iff this entry was spent on that date
     */
    public boolean wasSpentOn(LocalDate date) {
        return spent_on.equals(date);
    }

    /**
     * Checks if this entry was spent on a specific month
     *
     * @param month check spent with this month
     * @return true iff this entry was spent on that month
     */
    public boolean wasSpentOn(YearMonth month) {
        return YearMonth.from(spent_on).equals(month);
    }


    /**
     * @return the spent hours of this entry
     */
    public double getSpent() {
        return spent;
    }

    /**
     * @return the comment of this entry
     */
    public String getComment() {
        return comment;
    }

    /* ------------------------- properties setters ------------------------- */

    /**
     * @param comment new comment of this entry
     */
    public void setComment(String comment) {
        this.comment = comment;
    }

    /**
     * @param amount new hours to add to this entry (negative to subtract)
     */
    public void addSpent(double amount) {
        amount = Math.max(spent + amount, 0) - spent; // don't subtract what can't be substracted
        spent += amount;
        issue.addSpent(amount);
        assert issue.getSpent() >= 0;
    }

    /* ------------------------- uploading ------------------------- */

    /**
     * @return a list of changes made to this entry, as an object (empty means no changes)
     */
    public JSONObject getChanges() {
        JSONObject changes = new JSONObject();

        if (original == null || spent != original.getDouble("hours")) {
            // changed hours
            changes.put("hours", spent);
        }
        if (original == null || !Objects.equals(comment, original.optString("comments"))) {
            // changed comment
            changes.put("comments", comment);
        }
        if (id == RedmineManager.NONE) {
            // without original, this data is considered new
            changes.put("issue_id", issue.id);
            changes.put("spent_on", spent_on);
        }

        // return changes (empty for nothing)
        return changes;
    }

    /**
     * @return true if this entry requires upload, false otherwise
     */
    public boolean requiresUpload() {
        return !(
                getChanges().isEmpty() // no changes, no upload
                        ||
                        (this.id == RedmineManager.NONE && getSpent() <= 0) // no useful changes, no upload
        );
    }

    /**
     * Uploads an entry to redmine (unless not needed)
     *
     * @throws IOException on upload error
     */
    public void uploadTimeEntry() throws IOException {
        // get changes
        JSONObject changes = getChanges();

        // ignore unmodified
        if (changes.isEmpty()) return;

        if (id == RedmineManager.NONE) {
            if (getSpent() > 0) {
                // new entry with hours, create
                System.out.println("Creating entry with data: " + changes);
                if (RedmineManager.OFFLINE) return;
                if (UrlJSONKt.post(manager.domain + "time_entries.json?key=" + manager.key, new JSONObject().put("time_entry", changes)) != 201) {
                    throw new IOException("Error when creating entry with data: " + changes);
                }
            }
            //new entry without hours, ignore
        } else {
            if (getSpent() > 0) {
                // existing entry with hours, update
                System.out.println("Updating entry " + id + " with data: " + changes);
                if (RedmineManager.OFFLINE) return;
                if (UrlJSONKt.put(manager.domain + "time_entries/" + id + ".json?key=" + manager.key, new JSONObject().put("time_entry", changes)) != 200) {
                    throw new IOException("Error when updating entry " + id + " with data: " + changes);
                }
            } else {
                // existing entry without hours, delete
                System.out.println("Deleting entry " + id);
                if (RedmineManager.OFFLINE) return;
                if (UrlJSONKt.delete(manager.domain + "time_entries/" + id + ".json?key=" + manager.key) != 200) {
                    throw new IOException("Error when deleting entry " + id);
                }
            }
        }
    }
}

File: src/main/java/com/hiberus/anaya/redmineapi/TimeEntry.kt
package com.hiberus.anaya.redmineapi

import org.json.JSONObject
import java.time.LocalDate
import java.time.YearMonth

/**
 * 'You spent X hours into an issue with a message' object
 */
class TimeEntry {

    /* ------------------------- manager ------------------------- */

    private val remote: Remote // for remote interactions

    /* ------------------------- data ------------------------- */

    /**
     * the entry id in the database, null if this is a new entry without an associated id
     */
    val id: Int?

    /**
     * the issue id
     */
    var issue: Issue

    /**
     * date it was spent
     */
    var spent_on: LocalDate

    /**
     * the spent hours of this entry
     */
    var spent: Double
        private set

    /**
     * the comment of this entry
     */
    var comment: String

    /**
     * the original raw object, for diff purposes. Null if there is no original
     */
    private var original: JSONObject? = null


    /* ------------------------- constructors ------------------------- */

    /**
     * Creates a new entry from a json [rawEntry]
     * Issues must contain the issue!
     */
    internal constructor(rawEntry: JSONObject, issues: Iterable<Issue>, remote: Remote) {
        this.remote = remote
        original = rawEntry
        // creates a new entry from a json raw data
        id = rawEntry.getInt("id")
        issue = issues.first { issue: Issue -> issue.id == rawEntry.getIssueId() }
        spent_on = rawEntry.getSpentOnDate()
        spent = rawEntry.getDouble("hours")
        comment = rawEntry.optString("comments")
    }

    /**
     * Creates a new Time Entry for [issue] on [spent_on] with already [spent] hours and [comment]
     */
    internal constructor(issue: Issue, spent_on: LocalDate, spent: Double, comment: String, remote: Remote) {
        this.remote = remote
        id = null
        this.issue = issue
        this.spent_on = spent_on
        this.spent = spent
        this.comment = comment
    }

    /* ------------------------- checks ------------------------- */

    /**
     * Checks if this entry was spent on a specific date
     *
     * @param date check spent with this date
     * @return true iff this entry was spent on that date
     */
    fun wasSpentOn(date: LocalDate) = spent_on == date

    /**
     * Checks if this entry was spent on a specific month
     *
     * @param month check spent with this month
     * @return true iff this entry was spent on that month
     */
    fun wasSpentOn(month: YearMonth) = YearMonth.from(spent_on) == month

    /* ------------------------- modifiers ------------------------- */

    /**
     * @param amount new hours to add to this entry (negative to subtract)
     */
    fun addSpent(amount: Double) =
        // don't subtract what can't be subtracted
        ((spent + amount).coerceAtLeast(0.0) - spent).let {
            spent += it
            issue.addSpent(it)
        }

    /**
     * changes the spent hours to any [amount]
     */
    fun changeSpent(amount: Double) = addSpent(amount - spent)

    /* ------------------------- uploading ------------------------- */

    /**
     * a list of changes made to this entry, as an object (empty means no changes)
     */
    internal val changes
        get() = JSONObject().apply {
            if (spent != original?.getDouble("hours")) {
                // changed hours
                put("hours", spent)
            }
            if (comment != original?.optString("comments")) {
                // changed comment
                put("comments", comment)
            }
            if (spent_on != original?.getSpentOnDate()
                || id == null
            ) {
                // change date, or new entry
                put("spent_on", spent_on)
            }
            if (issue.id != original?.getIssueId()
                || id == null
            ) {
                // change issue, or new entry
                put("issue_id", issue.id)
            }
        }

    /**
     * iff this entry requires upload, false otherwise
     */
    val requiresUpload
        get() = !(changes.isEmpty // no changes, no upload
                || (id == null && spent <= 0)) // no useful changes, no upload

}

/* ------------------------- util ------------------------- */

internal fun JSONObject.getIssueId() = getJSONObject("issue").getInt("id") // returns the id from a rawEntry
internal fun JSONObject.getSpentOnDate() = LocalDate.parse(getString("spent_on")) // returns the spent_on date from a rawEntry
